packages/contracts-bedrock/contracts/periphery/TransferOnion.sol

```diff
@@ -0,0 +1,84 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.17;
+
+import { ReentrancyGuard } from "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import { ERC20 } from "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+/**
+ * @title  TransferOnion
+ * @notice TransferOnion is a hash onion for distributing tokens. The shell commits
+ *         to an ordered list of the token transfers and can be permissionlessly
+ *         unwrapped in order. The SENDER must `approve` this contract as
+ *         `transferFrom` is used to move the token balances.
+ */
+contract TransferOnion is ReentrancyGuard {
+    using SafeERC20 for ERC20;
```

Comment:
I assume that this will be used for the `GovernanceToken`? If so it's unnecessary since we know that it will throw on failure.

Reply:
Wow, this is expensive too!

Snapshot after removing:

```diff
-TransferOnionTest:test_constructor_succeeds() (gas: 563233)
-TransferOnionTest:test_unwrap_succeeds() (gas: 723651)
+TransferOnionTest:test_constructor_succeeds() (gas: 372457)
+TransferOnionTest:test_unwrap_succeeds() (gas: 531763)
```